lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/TaskRuntime.cs

```diff
@@ -92,50 +93,60 @@ public System.Threading.Tasks.Task RunTask()
 
             // Send heartbeat such that user receives a TaskRunning message.
             _currentStatus.SetRunning();
-            
+
             return System.Threading.Tasks.Task.Run(() =>
             {
                 Logger.Log(Level.Info, "Calling into user's task.");
                 return _userTask.Call(null);
-            }).ContinueWith(runTask =>
+            }).ContinueWith((System.Threading.Tasks.Task<byte[]> runTask) =>
+            {
+                try
                 {
-                    try
+                    // Task failed.
+                    if (runTask.IsFaulted)
                     {
-                        // Task failed.
-                        if (runTask.IsFaulted)
+                        if (runTask.Exception == null)
                         {
-                            Logger.Log(Level.Warning,
-                                string.Format(CultureInfo.InvariantCulture, "Task failed caused by exception [{0}]", runTask.Exception));
-                            _currentStatus.SetException(runTask.Exception);
-                            return;
+                            Logger.Log(Level.Error, "Task failed without an Exception.");
+                            _currentStatus.SetException(new ApplicationException());
                         }
-
-                        if (runTask.IsCanceled)
+                        else
                         {
-                            Logger.Log(Level.Warning,
-                                string.Format(CultureInfo.InvariantCulture, "Task failed caused by task cancellation"));
-                            return;
+                            var aggregateException = runTask.Exception.Flatten();
+                            _currentStatus.SetException(
+                                aggregateException.InnerExceptions.Count == 1 ?
+                                aggregateException.InnerExceptions.First() : aggregateException);
                         }
 
-                        // Task completed.
-                        var result = runTask.Result;
-                        Logger.Log(Level.Info, "Task Call Finished");
-                        _currentStatus.SetResult(result);
-                        if (result != null && result.Length > 0)
-                        {
-                            Logger.Log(Level.Info, "Task running result:\r\n" + System.Text.Encoding.Default.GetString(result));
-                        }
+                        return;
                     }
-                    finally
+
+                    if (runTask.IsCanceled)
```

Comment:
What could cause this cancelation? In this case, what shall we tall evaluator, is it Application error or system error?